### PR TITLE
Add recipes

### PR DESCRIPTION
**Pull request by jakubjen**

### CODE REVIEW

File: src/app/recipes/recipe-from/recipe-from.component.html [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+					formControlName="title"
+					[placeholder]="'Recipes.Title' | translate"
+					class="form-control"
+					[class.is-invalid]="
+						(title?.touched || ngForm.submitted) && !!title?.errors
+					"
+					id="title" />
+				<div
+					class="erros"
+					*ngIf="(title?.touched || ngForm.submitted) && !!title?.errors">
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.html:24]: why not just !recipeForm.controls?.get('name')?.valid

File: src/app/recipes/recipe-from/recipe-from.component.html [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+					</div>
+					<div
+						class="text-danger"
+						*ngIf="!!imageUrl?.errors?.['minlength'] || !!imageUrl?.errors?.['maxLength']">
+						<small>
+							{{
+								'App.Validation.Length'
+									| translate: { field: 'Image Url', min: 10, max: 100 }
+							}}
+						</small>
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.html:72]: think about creating a component for errors handling with
AbstractControl as an Input() and list all possible errors in this component

File: src/app/recipes/recipe-from/recipe-from.component.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+			Validators.required,
+			Validators.minLength(50),
+			Validators.maxLength(2000),
+		]),
+		time: new FormControl<number | null>(null, [Validators.required]),
+		ingredients: new FormArray<FormGroup>([]),
+		instructions: new FormArray<FormControl<string>>([]),
+	});
+
+	constructor() {}
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.ts:44]: remove unused constructor

File: src/app/recipes/recipe-from/recipe-from.component.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+	public handleSubmit(): void {
+		if (this.recipeForm.valid) {
+			this.submitRecipe.emit(this.recipeForm.value as Recipe);
+		}
+	}
+
+	get ingredients() {
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.ts:52]: return type

File: src/app/recipes/recipe-from/recipe-from.component.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		recipe.ingredients.forEach(() => this.addIngredient());
+		recipe.instructions.forEach(() => this.addInstruction());
+		this.recipeForm.patchValue(recipe);
+	}
+	@Output() submitRecipe = new EventEmitter<Recipe>();
+
+	public units = IngredientsUnit;
+
+	public recipeForm = new FormGroup({
+		title: new FormControl<string>('', [
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.ts:24]: i like your approach but this below is also working
```
title: ['', [ Validators.required,
		Validators.minLength(10),
		Validators.maxLength(200)]]
```

File: src/app/recipes/recipe-from/recipe-from.component.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+	get description() {
+		return this.recipeForm.get('description');
+	}
+
+	get time() {
+		return this.recipeForm.get('time');
+	}
+	get instructions() {
+		return this.recipeForm.get('instructions') as FormArray;
```
Comment (danieljanusm) [inline src/app/recipes/recipe-from/recipe-from.component.ts:85]: return types to all methods above

File: src/app/services/recipes/recipes.service.ts [hunk 7]
```diff
@@ -47,6 +47,12 @@ export class RecipesService {
 			});
 	}
 
+	public addRecipe(recipe: Recipe): void {
+		console.log('send');
```
Comment (danieljanusm) [inline src/app/services/recipes/recipes.service.ts:51]: console

**Review by danieljanusm — APPROVED** [review]

File: src/app/recipes/recipe-from/recipe-from.component.html (hunk 1 above)
Comment (jakubjen) [reply src/app/recipes/recipe-from/recipe-from.component.html:24]: I don't wont to user see error on start. In my opinion error should be visible after try of submitting form or when input lose focus.

Comment (danieljanusm) [reply src/app/recipes/recipe-from/recipe-from.component.html:24]: ok, error shouldn't be visible in this case when user didn't touch input, but if it's not working it's ok

File: src/app/recipes/recipe-from/recipe-from.component.html (hunk 2 above)
Comment (jakubjen) [reply src/app/recipes/recipe-from/recipe-from.component.html]: I create form error component

**Pull request merged**